---
title: "2.2_multiome_persisters"
author: "Anna Schwager"
date: "`r Sys.Date()`"
output:
  html_document:
    df_print: paged
    number_sections: yes
    toc: yes
    toc_float: yes
    code_folding: show
---

Load the global variables 
```{r source, message=F, warning=F, echo=F}
here::i_am("scripts/2.2_multiome_persisters.Rmd")
mainDir = here::here()
source(knitr::purl(file.path(mainDir,"scripts/global_variables.Rmd"), quiet=TRUE))
source(knitr::purl(file.path(mainDir, "scripts/functions.Rmd"), quiet=TRUE))

set.seed(123)
#scales::show_col(mypal)
```

# Preparation and object creation
## Setting directories and input files
```{r set_inputs, message=F, warning=F}

inputDir_local = file.path(inputDir, "hg38", "persisters", "one_cell_multiome")

outputDir_local = file.path(outputDir, "2.2_multiome_persisters") ; if(!file.exists(outputDir_local)){dir.create(outputDir_local)}
outputDir_objects = file.path(outputDir_local, "objects") ; if(!file.exists(outputDir_objects)){dir.create(outputDir_objects)}
outputDir_plots = file.path(outputDir_local, "plots") ; if(!file.exists(outputDir_plots)){dir.create(outputDir_plots)}
outputDir_tables = file.path(outputDir_local, "tables") ; if(!file.exists(outputDir_tables)){dir.create(outputDir_tables)}


ld <- list.dirs(inputDir_local)

fragpaths <- list.files(ld[grepl("/.*fragmentFiles", ld)], full.names = TRUE)
fragpaths <- fragpaths[grepl(".*/h3k4me1/.*fragmentFiles/.*tsv.gz$", fragpaths)]
names(fragpaths) <- c("h3k4me1")

rnapaths <- ld[grepl("/10XlikeMatrix_umi", ld)]

rm(ld)

```

## Loading annotation
```{r annot, message=F, warning=F}
consensus_peaks <- toGRanges(file.path(annotDir, "MM468_peaks_h3k4me1.bed"), format="BED", header=FALSE)
```

## Loading the bigwigs 
```{r bw_tab, message=F, warning=F}
ld <- list.dirs(inputDir)
bw_paths <- list.files(ld[grepl(".*/persisters/bigwig.*", ld)], full.names = TRUE)

names(bw_paths) <- c("pseudobulk", "max_cell")

list_rna_bw <- list(rna_all = bw_paths[["pseudobulk"]],
                    rna_max = bw_paths[["max_cell"]])

rm(ld)

```

## Loading the FACS data
```{r facs, message=F, warning=F}
ld <- list.dirs(inputDir_local)
facs_path <- list.files(ld[grepl("/facs", ld)], full.names = TRUE)
facs_data <- read.csv(facs_path)
rm(ld)

```

## Creating the object 
```{r create_object, message=F, warning=F, eval=F}

## loading RNA data
rna.data <- Read10X(data.dir = rnapaths)
rna_seurat <- CreateSeuratObject(counts = rna.data,
                             min.cells = 1,
                             min.features = -1,
                             project = "fresh_PDX")

## creating the multiome seurat objects 
seurat_list <- list()

for (i in names(fragpaths)){
    message(paste0("### Making fragment object for ", i))
    total_counts <- CountFragments(fragpaths[[i]])
    barcodes <- total_counts$CB
    frags <- CreateFragmentObject(path = fragpaths[[i]], cells = barcodes)
    
    message(paste0("### Making 50k bin matrix for ", i))
    bin50k_kmatrix = GenomeBinMatrix(
    frags,
    genome = hg38,
    cells = NULL,
    binsize = 50000,
    process_n = 10000,
    sep = c(":", "_"),
    verbose = TRUE)
    
    message(paste0("### Making peak matrix for ", i))
    peak_matrix = FeatureMatrix(
      frags,
      features = consensus_peaks,
      cells = NULL,
      sep = c("-", "-"),
      verbose = TRUE)
    
    message(paste0("### Creating chromatin assay for ", i))
    chrom_assay <- CreateChromatinAssay(
      counts = bin50k_kmatrix,
      sep = c(":", "_"),
      genome = "hg38",
      fragments = fragpaths[[i]],
      min.cells = 1,
      min.features = -1)
    
    message(paste0("### Creating Seurat object for ", i))
    seurat <- CreateSeuratObject(
      counts = chrom_assay,
      assay = "bin_50k")
    
    message(paste0("### Adding peak assay for ", i))
    seurat[["peaks"]] <- CreateChromatinAssay(
    counts = peak_matrix, genome = "hg38")
    2
    message(paste0("### Adding RNA assay for ", i))
    seurat <- AddMetaData(seurat, rna_seurat@meta.data)
    seurat[["RNA"]] <- rna_seurat@assays[["RNA"]]

    Annotation(seurat) <- annotations_hg38
    
    seurat <- AddMetaData(seurat, CountFragments(fragpaths[[i]]))
    seurat <- FRiP(object = seurat, assay = "peaks", total.fragments = "frequency_count")
    seurat@meta.data[["orig.ident"]] <- i
    seurat_list[[i]] <- seurat
    
    rm(seurat)
    rm(frags)
    rm(total_counts)
    rm(barcodes)
    rm(fragments_per_cell)
    rm(bin50k_matrix)
    rm(peak_matrix)
    rm(chromatin_assay)
}

seurat <- seurat_list[[1]]
saveRDS(seurat, file.path(outputDir_objects, "seurat_persisters_step1.rds"))

```

```{r reload1, message=F, warning=F}
seurat <- readRDS(file.path(outputDir_objects, "seurat_persisters_step1.rds"))
```


## FACS annotation
### Scatter plot
```{r facs1, message=F, warning=F, fig.show="hold", out.width="50%"}

  p_facs_log <- ggplot(facs_data, aes(x = log(H2Db), y = log(HLA))) +
  geom_pointdensity(adjust = 1, size = 1) +  # Use density-based coloring
  scale_color_viridis_c(option = "magma") +  # Magma colormap
  labs(x = "log Mouse H2Db", y = "log Human HLA", title = "Mouse H2Db vs Human HLA log") +
  theme_classic()

p_facs_log_density <- ggplot(facs_data, aes(x = log(H2Db), y = log(HLA))) +
  stat_density_2d(aes(fill = after_stat(level)), geom = "polygon", contour = TRUE, bins = 30, alpha = 0.2) +  
  scale_fill_viridis_c(option = "magma") +  # Magma colormap
  labs(x = "log Mouse H2Db", y = "log Human HLA", title = "Mouse H2Db vs Human HLA log") +
  theme_classic()

p_facs_log_density

```


```{r save_facs, message=F, warning=F, eval=F}

ggsave(file.path(outputDir_plots, "facs_persisters_density_sub.pdf"),
       plot = p_facs_log_density,
       device = "pdf",
       units = "mm",
       width = 100,
       height = 80)
```


### Adding FACS data as Seurat metadata
```{r add_facs_data, message=F, warning=F, fig.show="hold", out.width="50%"}

facs_persisters <- facs_data[facs_data$Annot == "Persister",]
row.names(facs_persisters) <- facs_persisters$CB

seurat <- AddMetaData(
  object = seurat,
  metadata = facs_persisters
)

```

# QC
## Weighted histogram
```{r qc1, message=F, warning=F, fig.show="hold", out.width="50%"}

plot_weighted_hist(seurat) + ggtitle("H3K4me1")
plot_weighted_hist(seurat, assay = "RNA") + ggtitle("RNA")
plot_weighted_hist(seurat, assay = "RNA_features") + ggtitle("RNA features")

```
## Filtering 
```{r filtering, message=F, warning=F }

min_reads_chromatin = 400
min_reads_rna = 1000

seurat@meta.data[["filtering"]] <- ifelse(seurat@meta.data$frequency_count > min_reads_chromatin & 
                                          seurat@meta.data$nCount_RNA > min_reads_rna ,
                                                                             'pass', 'fail')
```


## FrIP and N fragments plots

```{r qc2, message=F, warning=F, fig.show="hold", out.width="50%"}

p_count <- VlnPlot(subset(seurat, subset = filtering == "pass"), c("frequency_count"),
                   group.by = "orig.ident", split.by = NULL, pt.size = 0) +
                   labs(title = "Unique fragments per cell PDX") +
                   stat_summary(fun.y = median, geom='point', size = 2, colour = "black") +
                   theme(legend.position = "none") +
                   scale_fill_manual(values = c(mypal[2]))
print(p_count)

p_frip <- VlnPlot(subset(seurat, subset = filtering == "pass"), c("FRiP"),
                   group.by = "orig.ident", split.by = NULL, pt.size = 0) +
                   labs(title = "FRiP per cell PDX") +
                   stat_summary(fun.y = median, geom='point', size = 2, colour = "black") +
                   theme(legend.position = "none") +
                   scale_fill_manual(values = c(mypal[2]))
print(p_frip)

p_count_rna <- VlnPlot(subset(seurat, subset = filtering == "pass"), c("nCount_RNA"),
                       group.by = "orig.ident", split.by = NULL, pt.size = 0) +
                       labs(title = "Unique reads per cell PDX") +
                       stat_summary(fun.y = median, geom='point', size = 2, colour = "black") +
                       theme(legend.position = "none") +
                       scale_fill_manual(values = c(mypal[1]))
print(p_count_rna)

p_genes_rna <- VlnPlot(subset(seurat, subset = filtering == "pass"), c("nFeature_RNA"),
                        group.by = "orig.ident", split.by = NULL, pt.size = 0) +
                        labs(title = "Unique genes per cell PDX") +
                        stat_summary(fun.y = median, geom='point', size = 2, colour = "black") +
                        theme(legend.position = "none") +
                        scale_fill_manual(values = c(mypal[1]))
print(p_genes_rna)

```


```{r save1, message=F, warning=F, eval=F}

ggsave(file.path(outputDir_plots, "fragments_persisters.pdf"),
       plot = p_count,
       device = "pdf",
       units = "mm",
       width = 40,
       height = 90)

ggsave(file.path(outputDir_plots, "frip_persisters.pdf"),
       plot = p_frip,
       device = "pdf",
       units = "mm",
       width = 40,
       height = 90)

ggsave(file.path(outputDir_plots, "n_genes_persisters.pdf"),
       plot = p_genes_rna,
       device = "pdf",
       units = "mm",
       width = 40,
       height = 90)

ggsave(file.path(outputDir_plots, "rna_count_persisters.pdf"),
       plot = p_count_rna,
       device = "pdf",
       units = "mm",
       width = 40,
       height = 90)

```

## DNA to RNA reads
```{r qc_scatters, message=F, warning=F, fig.show="hold", out.width="50%"}
seurat <- subset(seurat, subset = filtering == "pass")

frags_to_rna_reads <- ggplot(seurat@meta.data, aes(x = frequency_count, y = nCount_RNA, color = orig.ident)) +
                             geom_point(alpha = 1, stroke = 0, size = 2) +  
                             labs(x = "N unique DNA fragments",
                                  y =  "N unique RNA reads",
                                title = "N unique DNA fragments vs. N unique RNA reads") +
                              theme_classic() +
                              scale_color_manual(values = c(mypal[2], mypal[3]))

frags_to_genes <- ggplot(seurat@meta.data, aes(x = frequency_count, y = nFeature_RNA, color = orig.ident)) +
                             geom_point(alpha = 1, stroke = 0, size = 2) +  
                             labs(x = "N unique DNA fragments",
                                  y =  "N unique genes",
                                title = "N unique genes vs. N unique RNA reads") +
                              theme_classic() +
                              scale_color_manual(values = c(mypal[2], mypal[2]))

genes_to_rna_reads <- ggplot(seurat@meta.data, aes(x = nFeature_RNA, y = nCount_RNA, color = orig.ident)) +
                             geom_point(alpha = 1, stroke = 0, size = 2) +  
                             labs(x = "N unique genes",
                                  y =  "N unique RNA reads",
                                title = "N unique genes vs. N unique RNA reads") +
                              theme_classic() +
                              scale_color_manual(values = c(mypal[2], mypal[2]))

frags_to_rna_reads
frags_to_genes
genes_to_rna_reads

```

```{r save2, message=F, warning=F, eval=F}
ggsave(file.path(outputDir_plots, "frags_to_rna_reads.pdf"),
       plot = frags_to_rna_reads,
       device = "pdf",
       units = "mm",
       width = 100,
       height = 80)

ggsave(file.path(outputDir_plots, "frags_to_genes.pdf"),
       plot = frags_to_genes,
       device = "pdf",
       units = "mm",
       width = 100,
       height = 80)

ggsave(file.path(outputDir_plots, "genes_to_rna_reads.pdf"),
       plot = genes_to_rna_reads,
       device = "pdf",
       units = "mm",
       width = 100,
       height = 80)
```

## Coverage plots

```{r prep_tracks, message=F, warning=F}
DefaultAssay(seurat) <- "bin_50k"
```

Sub-setting the max cell for single-cell tracks.
```{r subset_max, message=F, warning=F}
max_cell <- subset(seurat, frequency_count == max(seurat@meta.data$frequency_count))

```

### JUNB
```{r cov_plot_1, message=F, warning=F}

#chr19:12,782,200-12,814,051
roi <- GRanges(seqnames = Rle(c("chr19"), c(1)),
               ranges = IRanges(12782200, end = 12814051, names = "JUNB"))

gene_plot <- AnnotationPlot(
  object = seurat,
  region = roi)

cov_plot_k4 <- CoveragePlot(
  object = seurat,
  region = roi,
  annotation = FALSE,
  peaks = FALSE,
  window = 800) +
  ggtitle("H3K4me1 pseudobulk") + 
  scale_fill_manual(values = c(mypal[2]))

cov_plot_k4_max <- CoveragePlot(
  object = max_cell,
  region = roi,
  annotation = FALSE,
  peaks = FALSE,
  window = 800) +
  ggtitle("H3K4me1 max_cell") + 
  scale_fill_manual(values = c(mypal[2]))

cov_plot_rna <- BigwigTrack(
                     bigwig = list_rna_bw,
                     region = roi,
                     smooth = 200,
                     type = "coverage",
                     y_label = "Normalised signal") +
                     scale_fill_manual(values = rep(mypal[1], 2)) +
                     theme(legend.position = "none") + 
                     ggtitle("RNA pseudobulk")

p_junb <- CombineTracks(plotlist = list(cov_plot_rna, cov_plot_k4_max, cov_plot_k4, gene_plot),
                       heights = c(20,10,10,8)) &
                       theme(axis.title.y = element_text(size = 7))
p_junb

```
```{r save_tracks_1, message=F, warning=F}
ggsave(file.path(outputDir_plots, "junb_tracks.pdf"),
       plot = p_junb,
       device = "pdf",
       units = "mm",
       width = 200,
       height = 100)
```


```{r expression, message=F, warning=F}
DefaultAssay(seurat) <- "RNA"
p_junb_expression <- VlnPlot(seurat, c("JUNB", "PRDX2", "RNASEH2A"), layer = "counts", cols = rep(mypal[1], 3))

p_junb_expression

```
```{r save_3, message=F, warning=F}
ggsave(file.path(outputDir_plots, "junb_expression.pdf"),
       plot = p_junb_expression,
       device = "pdf",
       units = "mm",
       width = 120,
       height = 90)
```

